102 add md calcjob (#104)
stfc/aiida-mlip: docs/source/user_guide/calculations.rst

==============================
Calculations
==============================

> In these descriptions, we will assume that the code `janus-core <https://github.com/stfc/janus-core>`_ is installed and saved in the AiiDA database as an InstalledCode instance named 'janus@localhost'.

Reviewer: ```suggestion
In these examples, we will assume that the `janus-core <https://github.com/stfc/janus-core>`_ package is installed and saved in the AiiDA database as an `InstalledCode` instance named 'janus@localhost'.
```

> The structure file on which to run a calculation can be anything that `ASE <https://wiki.fysik.dtu.dk/ase/>`_ can read, indicated with the path `path/to/structure`.

Reviewer: What is indicated? Do you mean "It is specified by"?
```suggestion
The structure should be a path to a file. Here, the structure file is specified as `path/to/structure`.

.. note::
   Any format that `ASE <https://wiki.fysik.dtu.dk/ase/>`_ can read is a valid structure file for a calculation.
```

> The model file is a local file or a file to download, depending on the specific MLIP that is being used. In these examples, it is referred to as a local file `path/to/model`.

Reviewer: ```suggestion
The model file determines the specific MLIP to be used. I can be a local file or a URI to a file to download. In these examples, it is assumed to be a local file located at `path/to/model`.
```

==============================
Calculations
==============================

In these examples, we will assume that the `janus-core <https://github.com/stfc/janus-core>`_ package is installed and saved in the AiiDA database as an `InstalledCode` instance named 'janus@localhost'.

The structure should be a path to a file. Here, the structure file is specified as `path/to/structure`.

.. note::
   Any format that `ASE <https://wiki.fysik.dtu.dk/ase/>`_ can read is a valid structure file for a calculation.

> The model file determines the specific MLIP to be used. I can be a local file or a URI to a file to download. In these examples, it is assumed to be a local file located at `path/to/model`.

Reviewer 2: ```suggestion
The model file determines the specific MLIP to be used. It can be a local file or a URI to a file to download. In these examples, it is assumed to be a local file located at `path/to/model`.
```

.. code-block:: python

    SinglePointCalculation = CalculationFactory("janus.sp")

>     submit(SinglePointCalculation, code=InstalledCode, structure=StructureData, metadata="{"options": {"resources": {"num_machines": 1}}}")

Reviewer: This doesn't look right. Should those outer quotes be there?

    # Define calculation to run
    SinglePointCalculation = CalculationFactory("janus.sp")

    # Run calculation
    result, node = run_get_node(
        SinglePointCalculation,
        code=code,
        metadata=metadata,
        config=config,
    )

If a parameter is defined twice, in the config file and manually, the manually defined one will overwrite the config one.
If for example the same config file as before is used, but this time the parameter "struct" is added to the launch function, the code would look like this:

.. code-block:: python

    # Run calculation
    result, node = run_get_node(
        SinglePointCalculation,
        code=code,
        struct=StructureData(ase=read("path/to/structure2.xyz"))
        metadata=metadata,
        config=config,
    )

In this case  the structure used is going to be "path/to/structure2.xyz" rather than ""path/to/structure.cif", which was defined in the config file.

Refer to the API documentation for additional parameters that can be passed.
Some parameters are not required and don't have a default value set in aiida-mlip. In that case the default values will be the same as `janus <https://stfc.github.io/janus-core/>`_

> The only default parameters defined in aiida-mlip are the names of the input and output files, are they do not affect the results of the calculation itself, and are needed in AiiDA to parse the results.

Reviewer 2: ```suggestion
The only default parameters defined in aiida-mlip are the names of the input and output files, as they do not affect the results of the calculation itself, and are needed in AiiDA to parse the results.
```

.. code-block:: python

    SinglePointCalculation = CalculationFactory("janus.sp")
    submit(SinglePointCalculation, code=InstalledCode, structure=StructureData, metadata={"options": {"resources": {"num_machines": 1}}})

The inputs can be grouped into a dictionary:

.. code-block:: python

    inputs = {
            "metadata": {"options": {"resources": {"num_machines": 1}}},
            "code": InstalledCode,
            "architecture": Str,
            "structure": StructureData,
            "model": ModelData,
            "precision": Str,
            "device": Str,
        }
    SinglePointCalculation = CalculationFactory("janus.sp")
    submit(SinglePointCalculation, **inputs)

> Or they can be passed as a config file. The config file has to be structured as it would be for a janus calculation (refer to janus documentation) and passed as an AiiDA data type itself.

Reviewer 2: Maybe this should link to janus and/or the docs? I'll look for the rst syntax to suggest but you may know immediately?

> Or they can be passed as a config file. The config file has to be structured as it would be for a janus calculation (refer to `janus documentation <https://stfc.github.io/janus-core/apidoc/janus_core.html>`_ ) and passed as an AiiDA data type itself.

Reviewer 2: ```suggestion
Or they can be passed as a config file. The config file has to be structured as it would be for a janus calculation (refer to `janus documentation <https://stfc.github.io/janus-core/>`_ ) and passed as an AiiDA data type itself.
```
Optional but I'm thinking the main page might be better to link, since the API docs aren't the most user friendly starting point

Author: yeah I know, I put api cause at the moment the stuff is mostly there

    submit(GeomOptCalculation, code=InstalledCode, structure=StructureData, max_force=Float(0.1), vectors_only=Bool(True))

> As per the singlepoint calculation, the parameters can be provided in the form of a dictionary or a config file.

Reviewer: ```suggestion
.. note::

    As per the singlepoint calculation, the parameters can be provided as a dictionary or config file.
```

.. code-block:: python

    verdi run submit_geomopt.py "janus@localhost" --structure "path/to/structure" --model "path/to/model" --precision "float64" --device "cpu"

Molecular Dynamics calculation
------------------------------

> A `MD` Calculation represents a `Calcjob` object within the AiiDA framework.

Reviewer: ```suggestion
An `MD` Calculation represents a `Calcjob` object within the AiiDA framework.
```

Usage
^^^^^

This calculation can be executed using either the `run` or `submit` AiiDA commands.
Below is a usage example with some additional geometry optimisation parameters. These parameters must be AiiDA data types.

.. code-block:: python

    MDCalculation = CalculationFactory("janus.md")

>     submit(MDCalculation, code=InstalledCode, structure=StructureData, ensemble=Str("nve") md_dict=Dict({'temp':300,'steps': 4,'traj-every':3,'stats-every':1}))

Reviewer: ```suggestion
    submit(MDCalculation, code=InstalledCode, structure=StructureData, ensemble=Str("nve"), md_dict=Dict({'temp':300,'steps': 4,'traj-every':3,'stats-every':1}))
```

> As per the singlepoint calculation, the parameters can be provided in the form of a dictionary or a config file.

Reviewer: ```suggestion
.. note::

   As per the singlepoint calculation, the parameters can be provided as a dictionary or config file.
```
